Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,477 @@
+#include <wtf/Assertions.h>
+#include <wtf/Hasher.h>
+#include <wtf/glib/GRefPtr.h>
+#include <wtf/glib/GUniquePtr.h>
+#include <wtf/glib/WTFGType.h>
+
+static void webkit_image_gicon_interface_init(GIconIface*);
+static void webkit_image_gloadable_icon_interface_init(GLoadableIconIface*);
+
+struct _WebKitImagePrivate {
+    int width;
+    int height;
+    unsigned stride;
+    CompletionHandler<GRefPtr<GBytes>(WebKitImage*)> createBytesFn;
+
+    GRefPtr<GBytes> bytes;
+};
+
+/**
+ * WebKitImage:
+ *
+ * This class represents a WebKit image.
+ *
+ * At the moment, this assumes native images are in RGBA format, with 4 bytes per pixel.
+ * When instantiating this class, stride must be large enough to accommodate the image width times
+ * the bytes-per-pixel representation.
+ * When instantiating this class and providing a CompletionHandler to instantiate the bytes, make sure
+ * this uses #CompletionHandlerCallThread::AnyThread if you first instantiate the class bytes asynchronously.
+ *
+ * Since: 2.52
+ */
+WEBKIT_DEFINE_FINAL_TYPE_WITH_CODE(
+    WebKitImage, webkit_image, G_TYPE_OBJECT, GObject,
+    G_IMPLEMENT_INTERFACE(G_TYPE_ICON, webkit_image_gicon_interface_init)
+    G_IMPLEMENT_INTERFACE(G_TYPE_LOADABLE_ICON, webkit_image_gloadable_icon_interface_init))
+
+
+enum {
+    PROP_0,
+
+    PROP_WIDTH,
+    PROP_HEIGHT,
+    PROP_STRIDE,
+
+    N_PROPERTIES
+};
+
+enum class ImageFormatBPP : uint8_t {
+    RGBA8 = 4,
+};
+
+static std::array<GParamSpec*, N_PROPERTIES> sObjProperties;
+
+static void webkitImageSetProperty(GObject* object, guint propId, const GValue* value, GParamSpec* paramSpec)
+{
+    auto* image = WEBKIT_IMAGE(object);
+
+    switch (propId) {
+    case PROP_WIDTH:
+        image->priv->width = g_value_get_int(value);
+        break;
+    case PROP_HEIGHT:
+        image->priv->height = g_value_get_int(value);
+        break;
+    case PROP_STRIDE:
+        image->priv->stride = g_value_get_uint(value);
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
+    }
+}
+
+static void webkitImageGetProperty(GObject* object, guint propId, GValue* value, GParamSpec* paramSpec)
+{
+    auto* image = WEBKIT_IMAGE(object);
+
+    switch (propId) {
+    case PROP_WIDTH:
+        g_value_set_int(value, webkit_image_get_width(image));
+        break;
+    case PROP_HEIGHT:
+        g_value_set_int(value, webkit_image_get_height(image));
+        break;
+    case PROP_STRIDE:
+        g_value_set_uint(value, webkit_image_get_stride(image));
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
+    }
+}
+
+static void webkitImageConstructed(GObject* object)
+{
+    G_OBJECT_CLASS(webkit_image_parent_class)->constructed(object);
+
+    auto* image = WEBKIT_IMAGE(object);
+    auto* priv = image->priv;
+
+    gsize rowBytes = priv->width * static_cast<uint8_t>(ImageFormatBPP::RGBA8);
+    g_return_if_fail(priv->stride >= rowBytes);
```

Comment:
This is an internal check, so therefore it should be an assertion&mdash;the `g_return_[val_]if_fail()` macros are intended to be used at the top of public API functions.


```suggestion
    RELEASE_ASSERT(priv->width * RGBA8BytesPerPixel <= priv->stride);
```